File: src/BinarySearchTreeIterator/BSTIterator.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class BSTIterator {
    Stack<TreeNode> st = new Stack<>();
    public BSTIterator(TreeNode root) {
        pushLeftNodes(root);
    }

    private void pushLeftNodes(TreeNode node){
        while(node != null){
            st.push(node);
            node = node.left;
        }
    }
    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return !st.isEmpty();
    }

    /** @return the next smallest number */
    public int next() {
        TreeNode node = st.pop();
        int ret = node.val;
        pushLeftNodes(node.right);
        return ret;
    }
}

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = new BSTIterator(root);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: src/BinarySearchTreeIterator/BSTIterator.py
# Definition for a  binary tree node
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stack = []
        self.pushAll(self.stack, root)

    def pushAll(self, stack, node):
        while node:
            self.stack.append(node)
            node = node.left


    def hasNext(self):
        """
        :rtype: bool
        """
        return self.stack


    def next(self):
        """
        :rtype: int
        """
        node = self.stack.pop()
        self.pushAll(self.stack, node.right)
        return node.val


# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())